substrate/frame/data-market/src/types.rs

```diff
@@ -50,37 +57,78 @@ impl<BoundedString> VerificationProof<BoundedString> {
 	}
 }
 
+pub(crate) enum TransferFrom<T: Config> {
+	Origin(T::AccountId),
+	Escrow,
+}
+
 #[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
 #[cfg_attr(feature = "std", derive(Hash, Debug))]
-pub struct DataPurchaseRegisterDetails<AccountId, BlockNumber, Balance, AnyText> {
+pub struct DataDelegateContractDetail<AccountId, BlockNumber> {
+	pub data_owner: AccountId,
+	pub agency: AccountId,
+	pub data_owner_minimum_fee_ratio: u32,
+	pub deligated_data: AnyText,
+	pub effective_at: BlockNumber,
+	pub expired_at: BlockNumber,
+	pub signed_status: ContractSigner<AccountId>,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Hash, Debug))]
+pub struct DataDelegateContractParams<AccountId, BlockNumber> {
+	pub data_owner: AccountId,
+	pub data_owner_minimum_fee_ratio: u32,
+	pub deligated_data: AnyText,
+	pub duration: BlockNumber,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Hash, Debug))]
+pub struct DataPurchaseContractDetail<AccountId, BlockNumber, Balance> {
 	pub data_buyer: AccountId,
-	pub data_buyer_info: DataBuyerInfo<AnyText>,
+	pub data_verifier: Option<AccountId>,
+	pub effective_at: BlockNumber,
+	pub expired_at: BlockNumber,
+	pub data_purchase_info: DataPurchaseInfo<AnyText>,
+	pub system_token_id: u32,
+	pub agency: Option<AccountId>,
+	pub price_per_data: Balance,
+	pub deposit: Balance,
+	pub trade_count: Quantity,
+	pub data_trade_record: Vec<AccountId>,
+	pub signed_status: ContractSigner<AccountId>,
+}
```

Comment:
두 개를 이런식으로 통합할 수 있을거 같은데 어때요?

```suggestion
pub enum ContractType<AssetKind, Balance> {
	Purchase { asset: AssetKind, deposit: Balance },
	Delegate
}
pub struct ContractParty<Account>((Account, Account));
impl<Account> ContractParty<Account> {
	fn transfer() { .. }
}	
pub struct Contract<Account, BlockNumber, AssetKind, Balance> {
	pub contract_type: ContractType<AssetKind, Balance>,
	pub party: ContractParty<Account>,
	pub exp: BlockNumber,
	pub eff: BlockNumber,
	pub minimum_fee_ratio: u32
}
```

Reply:
각각의 Contract에 대해서 로직이나 Struct를 구성하는 데이터가 차이가 있는데 통합 할 필요가 있을까 싶습니다.

이렇게 구성하더라도 결국에 한쪽에만 있는 데이터들은 별도로 명시를 해 주어야 하는데 그러면 현재 나누어져 있는 구조랑 크게 다르지 않은 것 같습니다.

Reply:
로직은 `contract_type` 패턴 분리해서 구현한다면 서로 분리된 function 도 통합하여 구현할 수 있을거 같아서요. 필드값이 차이가 나지 않는것으로 보이고 이건 나중 task 로 잡아도 되지 않을까 싶긴합니다..